File: airbrakes/control_system/basic_kalman/kf.py
import numpy as np
from unittest import TestCase

class KF:
    def __init__(self, init_a, init_v, init_h, j_var):
        self.x = np.array([init_a, init_v, init_h])
        self.P = np.eye(3)
        self.j_var = j_var


    def predict(self, dt: float) -> None:
        F = np.array([[1, 0, 0], [dt, 1, 0], [0.5 * dt**2, dt, 1]])
        G = np.array([dt, 0.5 * dt**2, 1/6 * dt**3]).reshape(3, 1)

        new_x = F.dot(self.x)
        new_P = F.dot(self.P).dot(F.T) + G.dot(G.T) * self.j_var

        self.x = new_x
        self.P = new_P


    def update(self, meas_val, meas_var):
        z = np.array([meas_val])
        R = np.array([meas_var])
        H = np.array([1, 0, 0]).reshape((1, 3))

        y = z - H.dot(self.x)
        S = H.dot(self.P).dot(H.T) + R
        K = self.P.dot(H.T).dot(np.linalg.inv(S))

        new_x = self.x + K.dot(y)
        new_P = (np.eye(3) - K.dot(H)).dot(self.P)
        self.x = new_x
        self.P = new_P

    
    @property
    def h(self):
        return self.x[2]
    
    @property
    def v(self):
        return self.x[1]
    
    @property
    def a(self):
        return self.x[0]

File: airbrakes/control_system/archive/Cpp_model_test/kf.cpp
#include "kf.h"

KF::KF(float init_a, float init_v, float init_h, float j_var) {
    x << init_a, init_v, init_h;
    P = Eigen::Matrix3f::Identity();
    j = j_var;
}


void KF::predict(float dt) {
    Eigen::Matrix<float, 3, 3> F;
    Eigen::Matrix<float, 3, 1> G;

    F << 1.0, 0.0, 0.0,
            dt, 1.0, 0.0,
            1.0/2*dt*dt, dt, 1.0;
    G << dt, 1.0/2*dt*dt, 1.0/6*dt*dt*dt;

    x = F * x;
    P = F * P * F.transpose() + j * G * G.transpose();
}


void KF::update(float meas_val, float meas_var) {
    Eigen::Matrix<float, 1, 3> H;
    H << 1.0, 0.0, 0.0;

    float y = meas_val - H * x;
    float S = H * P * H.transpose() + meas_var;
    Eigen::Matrix<float, 3, 1> K;
    K = P * H.transpose() * 1.0/S;

    x = x + K * y;
    P = (Eigen::Matrix3f::Identity() - K*H) * P;
}